keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1176 @@
+With the removal of the `timeout` filter from the apiserver handler chain, the entire
+request handler will be executed on a single goroutine. Thus it will eliminate 
+the data race conditions that are inherent with the `timeout` filter.
+
+### Goals
+
+The primary goal is to reduce data race conditions by removing the `timeout`
+filter from the apiserver request handler chain.
+
+### Non-Goals
+
+- WATCH requests are outside the scope of this proposal, since currently we
+don't attach any deadline to the `Context` of a WATCH request. An orthogonal
+effort to allocate a deadline to the `Context`of WATCH request(s) is a
+prerequisite, and such effort is outside the scope of this proposal.
+
+- All other long-running requests are also outside the scope of this proposal.
+
+
+## Proposal
+
+A new feature flag `PerHandlerReadWriteTimeout` will be introduced, when enabled:
+- On start-up, the kube-apiserver will not add the [timeout filter](https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go#L38-L62) 
+  to the request handler chain.
+- To every incoming non long-running request, the kube-apiserver will apply
+  a request-scoped read/write deadline.
+- Use a synchronous finisher in place of the the [asynchronous finisher](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go#L87-L122) 
+  that is used today in order to asynchronously execute the admission and storage
+  handling of mutating request(s). On the other hand, the synchronous finisher 
+  will execute the request in the same goroutine.
+  
+We also need to maintain parity in terms of observability - the logs/metrics that
+are generated for requests that time out today, should also be available when this 
+feature is enabled.
+
+### User Stories (Optional)
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+
+### Risks and Mitigations
+
+##### Client Hanging Indefinitely:
+
+With the `timeout` filter enabled, it sends a `504 GatewayTimeout` response
+or an `error` to the client, if the request handler it is executing asynchronously
+does not return within the allotted deadline.
+
+If the client application does not enforce any client-side timeout, for example:
+ - not setting a timeout to the underlying `http.Client` or `http.Transport` object
+ - not setting a deadline to the `Context` object of an `http.Request`
+
+> It's worth noting that `client-go` does not enforce any global timeout by default.
+
+In this scenario, the client can potentially hang indefinitely, waiting for a
+response from the server. 
+It's worth noting that for HTTP/1x, [net/http does not execute the request handler in a new goroutine](https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/server.go#L2031-L2039).
```

Comment:
Is there some way for us to mimic the http2 behavior in http1 without using the `timeout` handler that we have today?